add PfsConfig.instStatusMask, updated tests accordingly

alefur opened the pull request.

PaulPrice commented on python/pfs/datamodel/pfsConfig.py:
```diff
@@ -1319,11 +1326,13 @@ def __init__(self, pfsDesignId, visit, raBoresight, decBoresight,
                  header=None,
-                 camMask=0):
+                 camMask=0,
+                 instStatusMask=0):
```
The default value is `INSROT_MISMATCH`? I think you need to add a `GOOD` status to `InstrumentStatusMask` as the `0` value.

PaulPrice commented on python/pfs/datamodel/pfsConfig.py:
```diff
@@ -1551,6 +1568,25 @@ def updateTargetPosition(self, ra, dec, pfiNominal, obstime, pfsUtilsVer):
+            raise ValueError(
+                f"Invalid instrument status flag: {mask}. Must be a valid InstrumentStatusMask enum.")
+
+        self.instStatusMask |= 1 << mask
```
`InstrumentStatusMask` inherits from `enum.IntEnum`, but it looks like you want something that inherits from `enum.IntFlag` since you want to be able to treat them as bitmasks. The current implementation doesn't allow the user to quickly identify what flags are set. For example:
```
>>> from pfs.datamodel.pfsConfig import DocEnum
>>> class MyMask(DocEnum):
...   NONE = 0, "None"
...   ONE = 1, "One"
...   TWO = 2, "Two"
... 
>>> mask = 0
>>> mask |= 1 << MyMask.ONE
>>> mask |= 1 << MyMask.TWO
>>> print(mask)
6
>>> print(MyMask(mask))
Traceback (most recent call last):
[...]
ValueError: 6 is not a valid MyMask
```

One option would be to provide a function that reports what bits are set when given an `instStatusMask` value from the pfsConfig. In that case, you should note in `InstrumentStatusMask` that the values provided are the bit numbers and not the mask (contrary to the name).
Another option would be to inherit from `IntFlag` (you'd want something similar to `DocEnum` layered over it too).
Another option would be to have an instance of `MaskHelper` instead of `InstrumentStatusMask`. The problem with that is that it doesn't have any place for descriptions of what the symbolic names mean.

PaulPrice reviewed (commented):
I think this needs some more thought. There's no way to quickly discover what the instrument status is given the `instStatusMask` value in the `pfsConfig`.

alefur commented:
Thank you for your feedback.
Yes, you were IntEnum was misleading, I've replaced with IntFlag. 
Unfortunately you can't put the descriptions directy in the class, so I added a InstrumentStatusDescription dictionary for that purpose. 
I also added functions to get and set the flags and get a human-readable version.

PaulPrice commented on python/pfs/datamodel/pfsConfig.py:
```diff
@@ -1551,6 +1574,41 @@ def updateTargetPosition(self, ra, dec, pfiNominal, obstime, pfsUtilsVer):
+        list of `InstrumentStatusFlag`
+            List of active flags in `instStatusMask`.
+        """
+        return [flag for flag in InstrumentStatusFlag if self.instStatusMask & flag]
```
```
return InstrumentStatusFlag(self.instStatusMask).name.split("|")
```

PaulPrice commented on python/pfs/datamodel/pfsConfig.py:
```diff
@@ -151,6 +151,17 @@ class FiberStatus(DocEnum):
     NOTCONVERGED = 7, "Cobra did not converge to the target."
 
 
+class InstrumentStatusFlag(enum.IntFlag):
```
This is now called `InstrumentStatusFlag`, while the pfsConfig member is called `instStatusMask`. Is the different deliberate or significant?

alefur replied:
InstrumentStatusFlag described the value/meaning of each bit 
whereas instStatusMask is the union of those bit.
I though it was different enough to keep separate name.
but you're right. perhaps just using instStatusFlag would be less confusing.

PaulPrice commented on python/pfs/datamodel/pfsConfig.py:
```diff
@@ -151,6 +151,17 @@ class FiberStatus(DocEnum):
     NOTCONVERGED = 7, "Cobra did not converge to the target."
 
 
+class InstrumentStatusFlag(enum.IntFlag):
```
For a bit more safety, I recommend:
```
@enum.verify(enum.UNIQUE)
class InstrumentStatusFlag(enum.IntFlag, boundary=enum.STRICT):
```
